src/parsers.ts

```diff
@@ -0,0 +1,55 @@
+export interface ValueNode {
+  type: 'value';
+  value: string|number|null;
+  unit: string|null;
+}
+
+/**
+ * Parses input strings that are a series of whitespace-separated CSS-like value
+ * expressions. Expressions in such strings include values such as:
+ *
+ *  - A color e.g., red or #8800ff
+ *  - A named orientation e.g., center
+ *  - A length e.g., 25px or 1m
+ *
+ * Some example value strings:
+ *
+ *  - 0 10px 100px
+ *  - red green 100%
+ *  - 180deg 3rad
+ *  - 1em
+ *
+ * NOTE(cdata): CSS function values currently not supported, so no rgb(...) etc.
+ */
+export const parseValues = (valuesString: string): Array<ValueNode> => {
+  return whitespaceSplit(valuesString.trim())
+      .map(valueString => parseAtomicValue(valueString));
+};
+
+const parseAtomicValue = (valueString: string): ValueNode => {
+  const value = extractValue(valueString);
+  const unit = (value != null ? valueString.replace(value, '') : null) || null;
+
+  return {type: 'value', value, unit};
+};
+
+const extractValue = (() => {
+  const VALUE_RE = /^(([^-^.^0-9].*)|([-]?[0-9.]+))/;
```

Comment:
This makes a bit more logical sense to me (ha, I know, regexes), although it still only extracts the values. I think it's a bit stricter (although still not strict) - specifically, "#a3#fd" now only matches "#a3".

`^((?:(?:#|[a-zA-Z])[a-zA-Z\d]*)|(?:-?[\d.]+))`

Oh! This will get the values too (in a second capture group) if my comment above makes sense, and then the whole string can be matched (taking advantage of the strictness).

`^((?:(?:#|[a-zA-Z])[a-zA-Z\d]*)|(?:-?[\d.]+))([a-zA-Z]*)$`

Reply:
Doing everything in one JavaScript `RegExp` match is way faster than approaches that require an extra `String.prototype.replace` (in Chrome, anyway):

![image](https://user-images.githubusercontent.com/240083/55367753-686eb900-54a3-11e9-94f2-7fddee944fb9.png)

Good call on this. Will fix.